lib/signex.ex

```diff
@@ -18,34 +18,66 @@ defmodule SignEx do
   end
 
 
-  def signature_valid?(
+  def signature_valid?(headers, params, public_key) do
+    case validate_signature(headers, params, public_key) do
+      {:ok, _} ->
+        true
+      {:error, _} ->
+        false
+    end
+  end
+  def validate_signature(
     headers,
     params = %SignEx.Parameters{algorithm: algorithm_str},
     public_key) when is_binary(public_key) and algorithm_str in @allowed_algorithms
   do
     algorithm = Algorithm.new(algorithm_str)
     case Base.decode64(params.signature) do
       {:ok, signature} ->
-        case fetch_keys(headers, params.headers) do
+        case Helper.fetch_keys(headers, params.headers) do
           {:ok, ordered_headers} ->
-            signing_string = compose_signing_string(ordered_headers)
-            :public_key.verify(signing_string, algorithm.digest, signature, decode_key(public_key))
-          {:error, _reason} ->
-            false
+            signing_string = Helper.compose_signing_string(ordered_headers)
+            :public_key.verify(signing_string, algorithm.digest, signature, Helper.decode_key(public_key))
+            |> if do
+              {:ok, :valid}
+            else
+              {:error, :signature_incorrect}
+            end
+          {:error, reason} ->
+            {:error, reason}
         end
       :error ->
-        false
+        {:error, :invalid_signature_encoding}
     end
   end
 
-  def verified?(body, metadata = %{}, params= %SignEx.Parameters{}, keystore) when
+
+  def verified?(body, metadata, params, keystore) do
+    case verify(body, metadata, params, keystore) do
+      {:ok, _} ->
+        true
+      {:error, _} ->
+        false
+    end
+  end
+  def verify(body, metadata = %{}, params= %SignEx.Parameters{}, keystore) when
       is_binary(body) do
-    with {:ok, public_key} <- fetch_key(keystore, params.key_id),
-         {:ok, full_digest} <- Map.fetch(metadata, "digest")
-    do
-      digest_valid?(body, full_digest) && signature_valid?(metadata, params, public_key)
-    else
-      _ -> false
+    case fetch_key(keystore, params.key_id) do
```

Comment:
I think it's not because we wanted to make sure every failure case returned a different error reason and this was a good way to make that explicit.